Update changelog (#653)
wasmx/fizzy: CHANGELOG.md

The format is based on [Keep a Changelog],
and this project adheres to [Semantic Versioning].

## [0.6.0] — Unreleased

Fizzy passes all of the official WebAssembly 1.0 tests. We are maintaining the WebAssembly 1.0 test suite with corrections and additions backported from the WebAssembly specification master branch. For this Fizzy release [the snapshot from 2020-11-13](https://github.com/wasmx/wasm-spec/tree/w3c-1.0-tests-backported-20201113/test/core) is used:
  - 18979 of 18979 binary parser and execution tests,
  - 999 of 999 validation tests,
  - 499 skipped due to testing text format parser.

> We continued working on performance improvements. Worth noting, the internal program representation has been changed to favorite execution time (8% faster) instead of instantiation time (9% slower).

Reviewer: ```suggestion
We continued working on performance improvements. Worth noting, the internal program representation has been changed to favour execution time (8% faster) instead of instantiation time (9% slower).
```

Author: Fixed.

The format is based on [Keep a Changelog],
and this project adheres to [Semantic Versioning].

## [0.6.0] — Unreleased

### Added

> - `fizzy-wasi` command-line tool to execute Wasm code using WASI interface 

Reviewer 2: Mention that WASI support is incomplete, but can be easily extended.

Author: Added, but maybe better to mention it above in the description of release.

      [#591](https://github.com/wasmx/fizzy/pull/591) [#595](https://github.com/wasmx/fizzy/pull/595)
    - Validating a module. [#538](https://github.com/wasmx/fizzy/pull/538)
    - Parsing, instantiation, (unsafe) execution. [#566](https://github.com/wasmx/fizzy/pull/566) 
- CMake package now can be used to integrate Fizzy. [#553](https://github.com/wasmx/fizzy/pull/553)
- New benchmarks. [#234](https://github.com/wasmx/fizzy/pull/234) [#632](https://github.com/wasmx/fizzy/pull/632)
- Support for building with libc++ on Linux by using CMake toolchain file. 
  [#597](https://github.com/wasmx/fizzy/pull/597)

### Changed

- Optimization in `execute`: now arguments are passed by a pointer (`const Value*`) instead of `span<const Value>`.
  [#552](https://github.com/wasmx/fizzy/pull/552) [#573](https://github.com/wasmx/fizzy/pull/573) 
  [#574](https://github.com/wasmx/fizzy/pull/574)
- Optimization in `fabs`, `fneg`, `fcopysign` instructions implementation. 
  [#560](https://github.com/wasmx/fizzy/pull/560)
- `popcnt` instructions implementation now uses standard library function when built with C++20. 
  [#551](https://github.com/wasmx/fizzy/pull/551) 

> - Optimization in pushing function result to the stack. [#602](https://github.com/wasmx/fizzy/pull/602)

Reviewer 2: This is not done yet.

- Optimization to store instruction immediate arguments together with instrustions in a single array. 
  [#618](https://github.com/wasmx/fizzy/pull/618) [#620](https://github.com/wasmx/fizzy/pull/620)
- Parsing now allocates `Module` dynamically and `std::unique_ptr<const Module>` is used in the API to represent a 
  module. [#575](https://github.com/wasmx/fizzy/pull/575) 
- Refactoring and simplification of table representation. [#616](https://github.com/wasmx/fizzy/pull/616)

> - Cmake project restructure: `fizzy::fizzy` library now includes only public headers, only `fizzy::fizzy-internal` 

Reviewer: ```suggestion
- CMake project restructuring: `fizzy::fizzy` library now includes only public headers, only `fizzy::fizzy-internal` 
```

  [#646](https://github.com/wasmx/fizzy/pull/646) [#649](https://github.com/wasmx/fizzy/pull/649)
- New unit tests. [#571](https://github.com/wasmx/fizzy/pull/571) [#580](https://github.com/wasmx/fizzy/pull/580) 
  [#641](https://github.com/wasmx/fizzy/pull/641) [#648](https://github.com/wasmx/fizzy/pull/648)
- Update Wasm spec test suite. [#534](https://github.com/wasmx/fizzy/pull/534)
- Test utils improvements. [#579](https://github.com/wasmx/fizzy/pull/579)
- Code and test cleanups and refactoring. [#527](https://github.com/wasmx/fizzy/pull/527) [#562](https://github.com/wasmx/fizzy/pull/562) 
  [#567](https://github.com/wasmx/fizzy/pull/567) [#578](https://github.com/wasmx/fizzy/pull/578) 
  [#606](https://github.com/wasmx/fizzy/pull/606) [#607](https://github.com/wasmx/fizzy/pull/607) 
  [#611](https://github.com/wasmx/fizzy/pull/611) [#617](https://github.com/wasmx/fizzy/pull/617)
  [#619](https://github.com/wasmx/fizzy/pull/619) [#623](https://github.com/wasmx/fizzy/pull/623)

### Fixed

- Remove unnecessary module copying when measuring instantiation performance for Fizzy in `fizzy-bench`. 
  [#581](https://github.com/wasmx/fizzy/pull/581)

> - Potential undefined behaviour for functions without locals. [#583](https://github.com/wasmx/fizzy/pull/583) 

Reviewer 2: This is not done yet.

The format is based on [Keep a Changelog],
and this project adheres to [Semantic Versioning].

## [0.6.0] — Unreleased

> Fizzy passes all of the [official test suite (spectest 1.0)](https://github.com/wasmx/wasm-spec/tree/w3c-1.0-tests-backported-20201113/test/core):

Reviewer 2: ```suggestion
Fizzy passes all of the official WebAssembly 1.0 tests. We are maintaining the WebAssembly 1.0 test suite with corrections and additions backported from the WebAssembly specification master branch. For this Fizzy release [the snapshot from 2020-11-13](https://github.com/wasmx/wasm-spec/tree/w3c-1.0-tests-backported-20201113/test/core) is used:
```

  - 18979 of 18979 binary parser and execution tests,
  - 999 of 999 validation tests,
  - 499 skipped due to testing text format parser.

### Added

> - `fizzy-wasi` command-line tool to execute Wasm code using WASI interface. (WASI support is not complete, yet, but can 

Reviewer 2: ```suggestion
- `fizzy-wasi` command-line tool to execute Wasm code using WASI interface. (WASI support is not complete yet, but can 
```

      [#631](https://github.com/wasmx/fizzy/pull/631) [#633](https://github.com/wasmx/fizzy/pull/633)
    - Getting type of a function. [#557](https://github.com/wasmx/fizzy/pull/557)
    - Getting module of an instance. [#599](https://github.com/wasmx/fizzy/pull/599)
    - Instantiation with resolving imported functions by names. [#608](https://github.com/wasmx/fizzy/pull/608)
    - `fizzy-bench` now includes `fizzyc` engine to benchmark Fizzy with C API. 
      [#561](https://github.com/wasmx/fizzy/pull/561)
- Rust bindings:
    - Building and crates.io package. [#584](https://github.com/wasmx/fizzy/pull/584) 
      [#585](https://github.com/wasmx/fizzy/pull/585) [#586](https://github.com/wasmx/fizzy/pull/586) 
      [#587](https://github.com/wasmx/fizzy/pull/587) [#589](https://github.com/wasmx/fizzy/pull/589)
      [#591](https://github.com/wasmx/fizzy/pull/591) [#595](https://github.com/wasmx/fizzy/pull/595)
    - Validating a module. [#538](https://github.com/wasmx/fizzy/pull/538)
    - Parsing, instantiation, (unsafe) execution. [#566](https://github.com/wasmx/fizzy/pull/566) 

> - CMake package now can be used to integrate Fizzy. [#553](https://github.com/wasmx/fizzy/pull/553)

Reviewer 2: ```suggestion
- CMake Package for easy Fizzy integration. [#553](https://github.com/wasmx/fizzy/pull/553)
```

- New benchmarks. [#234](https://github.com/wasmx/fizzy/pull/234) [#632](https://github.com/wasmx/fizzy/pull/632)
- Support for building with libc++ on Linux by using CMake toolchain file. 
  [#597](https://github.com/wasmx/fizzy/pull/597)

### Changed

> - Optimization in `execute`: now arguments are passed by a pointer (`const Value*`) instead of `span<const Value>`.

Reviewer 2: ```suggestion
- Optimization in `execute()`: now arguments are passed by a pointer (`const Value*`) instead of `span<const Value>`.
```

  [#552](https://github.com/wasmx/fizzy/pull/552) [#573](https://github.com/wasmx/fizzy/pull/573) 
  [#574](https://github.com/wasmx/fizzy/pull/574)
- Optimization in `fabs`, `fneg`, `fcopysign` instructions implementation. 
  [#560](https://github.com/wasmx/fizzy/pull/560)
- `popcnt` instructions implementation now uses standard library function when built with C++20. 
  [#551](https://github.com/wasmx/fizzy/pull/551) 
- Optimization to store instruction immediate arguments together with instrustions in a single array. 
  [#618](https://github.com/wasmx/fizzy/pull/618) [#620](https://github.com/wasmx/fizzy/pull/620)
- Parsing now allocates `Module` dynamically and `std::unique_ptr<const Module>` is used in the API to represent a 
  module. [#575](https://github.com/wasmx/fizzy/pull/575) 
- Refactoring and simplification of table representation. [#616](https://github.com/wasmx/fizzy/pull/616)

> - Cmake project restructuring: `fizzy::fizzy` library now includes only public headers, only `fizzy::fizzy-internal` 

Reviewer 2: ```suggestion
- CMake project restructuring: `fizzy::fizzy` library now includes only public headers, only `fizzy::fizzy-internal` 
```

  provides access to internal headers. [#550](https://github.com/wasmx/fizzy/pull/550)
- CI improvements. [#559](https://github.com/wasmx/fizzy/pull/559) [#590](https://github.com/wasmx/fizzy/pull/590)
  [#598](https://github.com/wasmx/fizzy/pull/598) [#600](https://github.com/wasmx/fizzy/pull/600) 
  [#603](https://github.com/wasmx/fizzy/pull/603) [#605](https://github.com/wasmx/fizzy/pull/605) 
  [#625](https://github.com/wasmx/fizzy/pull/625) [#629](https://github.com/wasmx/fizzy/pull/629)
  [#646](https://github.com/wasmx/fizzy/pull/646) [#649](https://github.com/wasmx/fizzy/pull/649)
- New unit tests. [#571](https://github.com/wasmx/fizzy/pull/571) [#580](https://github.com/wasmx/fizzy/pull/580) 
  [#641](https://github.com/wasmx/fizzy/pull/641) [#648](https://github.com/wasmx/fizzy/pull/648)

> - Update Wasm spec test suite. [#534](https://github.com/wasmx/fizzy/pull/534)

Reviewer 2: ```suggestion
- Update WebAssembly spec test suite. [#534](https://github.com/wasmx/fizzy/pull/534)
```
